libraries/plugins/es_objects/es_objects.cpp

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright (c) 2017 Cryptonomex, Inc., and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/es_objects/es_objects.hpp>
+
+//#include <graphene/app/impacted.hpp>
+
+//#include <graphene/chain/account_evaluator.hpp>
+//#include <graphene/chain/account_object.hpp>
+#include <graphene/chain/config.hpp>
+#include <graphene/chain/database.hpp>
+//#include <graphene/chain/evaluator.hpp>
+//#include <graphene/chain/operation_history_object.hpp>
+//#include <graphene/chain/transaction_evaluation_state.hpp>
+
+#include <fc/smart_ref_impl.hpp>
+#include <fc/thread/thread.hpp>
+
+#include <curl/curl.h>
+#include <boost/algorithm/string/replace.hpp>
+#include <graphene/chain/proposal_object.hpp>
+#include <boost/algorithm/string/join.hpp>
+#include <graphene/chain/balance_object.hpp>
+
+
+namespace graphene { namespace es_objects {
+
+namespace detail
+{
+
+class es_objects_plugin_impl
+{
+   public:
+      es_objects_plugin_impl(es_objects_plugin& _plugin)
+         : _self( _plugin )
+      {  curl = curl_easy_init(); }
+      virtual ~es_objects_plugin_impl();
+
+      void updateDatabase( const vector<object_id_type>& ids , bool isNew);
+
+      graphene::chain::database& database()
+      {
+         return _self.database();
+      }
+
+      es_objects_plugin& _self;
+      std::string _es_objects_elasticsearch_url = "http://localhost:9200/";
+      uint32_t _es_objects_bulk_replay = 1000;
+      uint32_t _es_objects_bulk_sync = 2;
+      bool _es_objects_proposals = true;
+      bool _es_objects_accounts = true;
+      bool _es_objects_assets = true;
+      bool _es_objects_balances = true;
+      bool _es_objects_logs = true;
+      CURL *curl; // curl handler
```

Comment:
Do we want to use a new curl handler for each plugin?

Reply:
interesting question, this is debatable at least to me. using the same handler for all requests that goes to the same url is recommended for performance and doing otherwise will be impossible for our purposes in the original plugin(connect and disconnect with each request will be just too slow in a replay).

i made a new handler here because i am not sure how much is too much when they run at the same time. the performance of sending bulks in 5k operation batches were proven to be good enough in a replay scenario while if we add more queries to the same handler we could run into performance issues with more data coming in by new plugins trying to store more data in the database.  for example if first plugin is sending a bulk of 5k operations(10000 lines of text ) and this new plugin is trying to send 1k lines more it will need to wait until the handler is available to do it ending in some delay that i am not sure if it will be a problem in practice. 

basically for that reason a new handler was added and answering the question; yes. the intention was to use a new handler with each plugin so they can be sending data in "parallel". 

i was able to get in sync a node with the 2 plugins running at the same time and it was ok this way, i can make some tests by using the same handler and check that if you think it worth it.

Reply:
Actually I'm not sure, so was asking.
If the 2 plugins are merged into one, there will be only one handler by nature.